### PR TITLE
SQL: Enable accurate hit tracking on demand

### LINKED ISSUE #33288: SQL: disable track total hits
In the vast majority of cases, SQL doesn't need to track the total amount of hits - as such it should disable it to optimize the query execution (especially in the case of `composite`).

### LINKED ISSUE #37971: SQL: Consider relaxing tracking total hits count
With the addition of approximate vs accurate total hit counts, we should consider whether we could relax this requirements for queries that do not need this.
Such as queries that do not have any count aggregation (which might be computed automatically).
Related to https://github.com/elastic/elasticsearch/issues/33288

### PR DESCRIPTION
**Pull request by costin**
Queries that require counting of all hits (COUNT(*) on implicit
group by), now enable accurate hit tracking.

Fix #37971
Closes https://github.com/elastic/elasticsearch/issues/33288

### CODE REVIEW

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java [hunk 1]
```diff
@@ -689,4 +689,40 @@ public void testTopHitsAggregationWithTwoArgs() {
...
+        EsQueryExec eqe = (EsQueryExec) p;
+        assertFalse("Should NOT be tracking hits", eqe.queryContainer().shouldTrackHits());
+    }
+
+    public void testNoCountDoesNotTrackHits() throws Exception {
+        PhysicalPlan p = optimizeAndPlan("SELECT int FROM test");
+        assertEquals(EsQueryExec.class, p.getClass());
+        EsQueryExec eqe = (EsQueryExec) p;
+        assertFalse("Should NOT be tracking hits", eqe.queryContainer().shouldTrackHits());
+    }
```
Comment (astefan) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java]: The only test missing I think it's with `COUNT(ALL field_name)`.

**Review by astefan — APPROVED** [review]
LGTM. Left one comment about a missing test case.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java [hunk 2]
```diff
@@ -391,11 +392,15 @@ else if (child instanceof GroupingFunction) {
...
                 Count c = (Count) f;
                 // COUNT(*) or COUNT(<literal>)
-                if (c.field() instanceof Literal) {
+                if (c.field().foldable()) {
                     AggRef ref = groupingAgg == null ?
                             GlobalCountRef.INSTANCE :
                             new GroupByRef(groupingAgg.id(), Property.COUNT, null);
 
+                    // if the count points to the total track hits, enable accurate count retrieval
+                    if (ref == GlobalCountRef.INSTANCE) {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java]: I would add this to an if block in combination with the above check on `groupingAgg == null`.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/container/QueryContainer.java [hunk 3]
```diff
@@ -81,23 +80,18 @@
...
 
-    public QueryContainer(Query query,
-            Aggs aggs,
-            List<Tuple<FieldExtraction, ExpressionId>> fields,
-            AttributeMap<Attribute> aliases,
-            Map<String, GroupByKey> pseudoFunctions,
-            AttributeMap<Pipe> scalarFunctions,
-            Set<Sort> sort,
-            int limit) {
+    public QueryContainer(Query query, Aggs aggs, List<Tuple<FieldExtraction, ExpressionId>> fields, AttributeMap<Attribute> aliases,
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/container/QueryContainer.java]: I don't know if it's the checkstyle that we follow, but personally I prefer the one arg per line for method signatures when the don't fit in one line, as it makes it more readable imho.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java [hunk 4]
```diff
@@ -689,4 +689,40 @@ public void testTopHitsAggregationWithTwoArgs() {
...
+
+
+    public void testGlobalCountInImplicitGroupByForcesTrackHits() throws Exception {
+        PhysicalPlan p = optimizeAndPlan("SELECT COUNT(*) FROM test");
+        assertEquals(EsQueryExec.class, p.getClass());
+        EsQueryExec eqe = (EsQueryExec) p;
+        assertTrue("Should be tracking hits", eqe.queryContainer().shouldTrackHits());
+    }
+
+    public void testGlobalCountInSpecificGroupByDoesNotForcesTrackHits() throws Exception {
```
Comment (matriv) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java]: typo: Forces -> Force

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java [hunk 5]
```diff
@@ -689,4 +689,40 @@ public void testTopHitsAggregationWithTwoArgs() {
...
+        PhysicalPlan p = optimizeAndPlan("SELECT COUNT(*) FROM test GROUP BY int");
+        assertEquals(EsQueryExec.class, p.getClass());
+        EsQueryExec eqe = (EsQueryExec) p;
+        assertFalse("Should NOT be tracking hits", eqe.queryContainer().shouldTrackHits());
+    }
+
+    public void testDistinctCountDoesNotTrackHits() throws Exception {
```
Comment (matriv) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java]: The method name is wrong and should be swapped with the one from the next one.

**Review by matriv — APPROVED** [review]
LGTM. Left some comments though, mostly minor.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/container/QueryContainer.java (hunk 3 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/container/QueryContainer.java]: Auto-formatting...

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java (hunk 1 above)
Comment (costin) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java]: That's an alias for `COUNT(field_name)` so it should just work however I've added one nevertheless.

**Pull request merged**